bundles/org.openhab.binding.ipcamera/src/main/java/org/openhab/binding/ipcamera/handler/IpCameraHandler.java

```diff
@@ -0,0 +1,2020 @@
+                }
+            }
+            return true;
+        } else {
+            cameraConfigError(
+                    "Camera is asking for Basic Auth when you have not provided a username and/or password !");
+        }
+        return false;
+    }
+
+    private String getCorrectUrlFormat(String longUrl) {
+        String temp = longUrl;
+        URL url;
+
+        try {
+            url = new URL(longUrl);
+            int port = url.getPort();
+            if (port == -1) {
+                if (url.getQuery() == null) {
+                    temp = url.getPath();
+                } else {
+                    temp = url.getPath() + "?" + url.getQuery();
+                }
+            } else {
+                if (url.getQuery() == null) {
+                    temp = ":" + url.getPort() + url.getPath();
+                } else {
+                    temp = ":" + url.getPort() + url.getPath() + "?" + url.getQuery();
+                }
+            }
+        } catch (MalformedURLException e) {
+            if (!longUrl.equals("ffmpeg")) {
+                cameraConfigError("A non valid URL has been given to the binding, check they work in a browser.");
+                return "";
+            }
+        }
+        return temp;
+    }
+
+    private void cleanChannels() {
+        lock.lock();
+        for (byte index = 0; index < listOfRequests.size(); index++) {
+            logger.debug("Channel status is {} for URL:{}", listOfChStatus.get(index), listOfRequests.get(index));
+            switch (listOfChStatus.get(index)) {
+                case 2: // Open and OK to reuse
+                case 1: // Open
+                case 0: // Closing but still open
+                    Channel channel = listOfChannels.get(index);
+                    if (channel.isOpen()) {
+                        break;
+                    } else {
+                        listOfChStatus.set(index, (byte) -1);
+                        logger.warn("Cleaning the channels has just found a connection with wrong open state.");
+                    }
+                case -1: // closed
+                    listOfRequests.remove(index);
+                    listOfChStatus.remove(index);
+                    listOfChannels.remove(index);
+                    listOfReplies.remove(index);
+                    index--;
+                    break;
+            }
+        }
+        lock.unlock();
+    }
+
+    private void closeChannel(String url) {
+        lock.lock();
+        try {
+            for (byte index = 0; index < listOfRequests.size(); index++) {
+                if (listOfRequests.get(index).equals(url)) {
+                    switch (listOfChStatus.get(index)) {
+                        case 2: // Still open and OK to reuse
+                        case 1: // Still open
+                        case 0: // Marked as closing but channel still needs to be closed.
+                            Channel chan = listOfChannels.get(index);
```

Comment:
Agree, I was planning on doing that after discussing this with a friend who is a programmer that helps when I get stuck. The code has been working so I have left it alone, but to make it easier to modify in the future the class is the better way and it will solve the next few concerns you have regarding the unlock.

Reply:
Done, created a new class and its much nicer and easier to change the binding now without risk of breaking.